### PR TITLE
When casting resource.Quantity to int64 it may overflow

### PR DESCRIPTION
**Pull request by gnufied**
We need to ensure that when converting resource.Quantity
to int64 it does not overflow and if it does, then an error is
returned back to the user rather than attempting expansion and
provisioning operations with scaled value.

This is a version of PR from https://github.com/kubernetes/kubernetes/pull/67969/files

```release-note
Prevent PVC requested size overflow when expanding or creating a volume
```

cc @wongma7 
/sig storage

### CODE REVIEW

File: staging/src/k8s.io/cloud-provider/volume/helpers/rounding.go [hunk 1]
```diff
@@ -39,74 +39,88 @@ const (
...
 	return roundUpSizeInt(requestBytes, GiB)
 }
 
 // RoundUpToMBInt rounds up given quantity to chunks of MB. It returns an
 // int instead of an int64 and an error if there's overflow
 func RoundUpToMBInt(size resource.Quantity) (int, error) {
-	requestBytes := size.Value()
+	requestBytes, ok := size.AsInt64()
+	if !ok {
+		return 0, fmt.Errorf("quantity %v is too gerat, overflows int64", size)
```
Comment (wongma7) [inline staging/src/k8s.io/cloud-provider/volume/helpers/rounding.go:102]: typo, s/gerat/great/

File: staging/src/k8s.io/cloud-provider/volume/helpers/rounding_test.go [hunk 2]
```diff
@@ -22,61 +22,12 @@ import (
...
-
 func Test_RoundUpToGiB(t *testing.T) {
 	testcases := []struct {
-		name       string
-		resource   resource.Quantity
-		roundedVal int64
+		name         string
+		resource     resource.Quantity
+		roundedVal   int64
+		expecteError bool
```
Comment (wongma7) [inline staging/src/k8s.io/cloud-provider/volume/helpers/rounding_test.go:30]: typo, expectError or expectedError

File: pkg/volume/portworx/portworx_util.go [hunk 3]
```diff
@@ -211,7 +214,11 @@ func (util *portworxVolumeUtil) ResizeVolume(spec *volume.Spec, newSize resource
 	}
 
 	vol := vols[0]
-	newSizeInBytes := uint64(volumehelpers.RoundUpToGiB(newSize) * volumehelpers.GiB)
+	requestGiB, err := volumehelpers.RoundUpToGiB(newSize)
+	if err != nil {
+		return err
+	}
+	newSizeInBytes := uint64(requestGiB * volumehelpers.GiB)
```
Comment (wongma7) [inline pkg/volume/portworx/portworx_util.go:221]: could requestGiB * volumehelpers.GiB overflow?

File: staging/src/k8s.io/cloud-provider/volume/helpers/rounding.go (hunk 1 above)
Comment (wongma7) [reply staging/src/k8s.io/cloud-provider/volume/helpers/rounding.go:102]: haha the linter caught this. 

```
staging/src/k8s.io/cloud-provider/volume/helpers/rounding.go:46:43: "gerat" is a misspelling of "great"
```

File: pkg/volume/portworx/portworx_util.go (hunk 3 above)
Comment (wongma7) [reply pkg/volume/portworx/portworx_util.go:221]: I suppose here it just needs to do something like 

newSizeInBytes, ok := newSize.AsUint64

assuming  AsUint64 function exists, if not then uint64(newSize.AsInt64)

File: staging/src/k8s.io/cloud-provider/volume/helpers/rounding.go (hunk 1 above)
Comment (gnufied) [reply staging/src/k8s.io/cloud-provider/volume/helpers/rounding.go:102]: yep. I need a git hook (that does not run forever!) for this stuff. :-)

File: pkg/volume/portworx/portworx_util.go (hunk 3 above)
Comment (gnufied) [reply pkg/volume/portworx/portworx_util.go]: yeah that is what I ended up doing.

**Comment by wongma7** [comment]
/lgtm

thank you for picking this up!!

File: pkg/volume/scaleio/sio_volume_test.go [hunk 4]
```diff
@@ -425,54 +425,6 @@ func TestVolumeProvisionerWithIncompleteConfig(t *testing.T) {
 	}
 }
 
-func TestVolumeProvisionerWithZeroCapacity(t *testing.T) {
```
Comment (andrewsykim) [inline pkg/volume/scaleio/sio_volume_test.go:428]: Update this test to validate the minimum capacity is being used?

File: staging/src/k8s.io/cloud-provider/volume/helpers/rounding_test.go [hunk 5]
```diff
@@ -162,7 +127,7 @@ func Test_RoundUpToMB(t *testing.T) {
 
 	for _, test := range testcases {
 		t.Run(test.name, func(t *testing.T) {
-			val := RoundUpToMB(test.resource)
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/volume/helpers/rounding_test.go:165]: Can we add an overflow test case for the other rounding functions?

Comment (gnufied) [reply staging/src/k8s.io/cloud-provider/volume/helpers/rounding_test.go:165]: fixed

File: pkg/volume/scaleio/sio_volume_test.go (hunk 4 above)
Comment (gnufied) [reply pkg/volume/scaleio/sio_volume_test.go:428]: fixed

File: pkg/volume/scaleio/sio_volume_test.go [hunk 6]
```diff
@@ -425,8 +426,8 @@ func TestVolumeProvisionerWithIncompleteConfig(t *testing.T) {
...
-	plugMgr, tmpDir := newPluginMgr(t, makeScaleIOSecret(testSecret, testns))
+func TestVolumeProvisionerWithSecretNamespace(t *testing.T) {
```
Comment (andrewsykim) [inline pkg/volume/scaleio/sio_volume_test.go:429]: This change seems unrelated? Maybe pulled in from another branch by accident?

File: staging/src/k8s.io/cloud-provider/volume/helpers/rounding.go [hunk 7]
```diff
@@ -117,6 +131,16 @@ func RoundUpToKiBInt(size resource.Quantity) (int, error) {
 	return roundUpSizeInt(requestBytes, KiB)
 }
 
+// RoundUpToGiBInt32 rounds up given quantity up to chunks of GiB. It returns an
+// int32 instead of an int64 and an error if there's overflow
+func RoundUpToGiBInt32(size resource.Quantity) (int32, error) {
```
Comment (andrewsykim) [inline staging/src/k8s.io/cloud-provider/volume/helpers/rounding.go:136]: I don't think there's a test for this one, can we add one?

File: pkg/volume/scaleio/sio_volume_test.go (hunk 6 above)
Comment (gnufied) [reply pkg/volume/scaleio/sio_volume_test.go]: hmm weird. It does. I fixed it.

File: staging/src/k8s.io/cloud-provider/volume/helpers/rounding.go (hunk 7 above)
Comment (gnufied) [reply staging/src/k8s.io/cloud-provider/volume/helpers/rounding.go:136]: Added.

**Review by andrewsykim — COMMENTED** [review]
/approve
/lgtm

**Comment by andrewsykim** [comment]
/lgtm cancel

Need to update bazel

**Comment by gnufied** [comment]
@andrewsykim sorry about that. Fixed.

**Review by andrewsykim — COMMENTED** [review]
/approve
/lgtm

**Comment by gnufied** [comment]
Failing test look like regular culprits unrelated to change. 

/test pull-kubernetes-integration

**Comment by gnufied** [comment]
Filed https://github.com/kubernetes/kubernetes/issues/91075 for flaking integration test

**Review by andrewsykim — COMMENTED** [review]
/lgtm
